### DISCUSSION TITLE
Class data cases do not always show in Test Explorer

### DISCUSSION THREAD

**Discussion by jchesshir** [xunit/xunit#2350] [Question] [answered]
I'm having an issue, but I can't explain why it is happening sometimes and other times not.  When I make this kind of class data...

```
		private class TestClassGoodData : ArrayList, IEnumerable<object[]>
		{
			new public IEnumerator<object[]> GetEnumerator()
			{
				yield return new object[] { "Random Text", false, null};
				yield return new object[] { "LASTACCOUNTINGPERIOD", true, new DateTime(2021, 7, 8) };
				yield return new object[] { "NEXTACCOUNTINGPERIOD", true, new DateTime(2021, 8, 10)  };
			}
		}
```
...Visual Studio 2019's Text Explorer is able to enumerate the test cases directly, like this:
![image](https://user-images.githubusercontent.com/11528004/125312168-a9e4a480-e2f9-11eb-905c-39a514c8cf44.png)

But this kind of class data...
```
		private class TestDataWithCustomObject : ArrayList, IEnumerable<object[]>
		{
			new public IEnumerator<object[]> GetEnumerator()
			{
				yield return new object[] { null, false, null};
				yield return new object[] { "test", false, null };
				yield return new object[] { "badTEst, 45", false, null };
				yield return new object[] { "45, Bad Test2", false, null };
				yield return new object[] { "Good Test", true, new CustomClass() {Latitude = 45, Longitude = 35}};
			}
		}
```
...causes VS to display only a single test case.  It only shows the individual test cases in the summary:
![image](https://user-images.githubusercontent.com/11528004/125313084-83733900-e2fa-11eb-8b99-c277d1b92747.png)

Is this another Visual Studio bug, or is there something I can do to get the CustomClass to help XUnit provide test case details to Visual Studio?

**Comment by bradwilson** [comment, marked as answer]
You need CustomClass to support xUnit.net's serialization by implementing [`IXunitSerializable`](https://github.com/xunit/abstractions.xunit/blob/main/src/xunit.abstractions/Serialization/IXunitSerializable.cs).

**Reply by jchesshir** [reply]
I found this article on [how to use IXunitSerializable](https://darchuk.net/2019/04/12/serializing-xunit-test-cases/).  However, after it finishes compiling, Text Explorer seems to just hang trying to load the individual test cases.  The test case that uses the Data Object displays, but none of the individual test cases show.  Any idea what I might be doing wrong?

Currently running:
> Microsoft Visual Studio Professional 2019
> Version 16.10.3

**Reply by bradwilson** [reply]
There are two substantial bugs in that sample:

- The Serialize implementation should be serializing `IsValid`
- The Deserialize implementation is completely missing

In addition, the sample is incomplete because it's missing the production code that's under test (so it's not simple to just copy/paste it and try it out). So it's impossible to know what's causing your hang. Can you provide a full repro project?
